scanpy/external/pp/_hashsolo.py

```diff
@@ -0,0 +1,388 @@
+    )
+    num_of_noise_barcodes = num_of_barcodes - number_of_non_noise_barcodes
+
+    # assume log normal
+    data = np.log(data + 1)
+    data_arg = np.argsort(data, axis=1)
+    data_sort = np.sort(data, axis=1)
+
+    # global signal and noise counts useful for when we have few cells
+    # barcodes with the highest number of counts are assumed to be a true signal
+    # barcodes with rank < k are considered to be noise
+    global_signal_counts = np.ravel(data_sort[:, -1])
+    global_noise_counts = np.ravel(data_sort[:, :-number_of_non_noise_barcodes])
+    global_mu_signal_o, global_sigma_signal_o = np.mean(global_signal_counts), np.std(
+        global_signal_counts
+    )
+    global_mu_noise_o, global_sigma_noise_o = np.mean(global_noise_counts), np.std(
+        global_noise_counts
+    )
+
+    noise_params_dict = {}
+    signal_params_dict = {}
+
+    # for each barcode get  empirical noise and signal distribution parameterization
+    for x in np.arange(num_of_barcodes):
+        sample_barcodes = data[:, x]
+        sample_barcodes_noise_idx = np.where(data_arg[:, :num_of_noise_barcodes] == x)[
+            0
+        ]
+        sample_barcodes_signal_idx = np.where(data_arg[:, -1] == x)
+
+        # get noise and signal counts
+        noise_counts = sample_barcodes[sample_barcodes_noise_idx]
+        signal_counts = sample_barcodes[sample_barcodes_signal_idx]
+
+        # get parameters of distribution, assuming lognormal do update from global values
+        noise_param = gaussian_updates(
+            noise_counts, global_mu_noise_o, global_sigma_noise_o
+        )
+        signal_param = gaussian_updates(
+            signal_counts, global_mu_signal_o, global_sigma_signal_o
+        )
+        noise_params_dict[x] = noise_param
+        signal_params_dict[x] = signal_param
+
+    counter_to_barcode_combo = {}
+    counter = 0
+
+    # for each combination of noise and signal barcode calculate probiltiy of in silico and real cell hypotheses
+    for noise_sample_idx, signal_sample_idx in product(
+        np.arange(num_of_barcodes), np.arange(num_of_barcodes)
+    ):
+        signal_subset = data_arg[:, -1] == signal_sample_idx
+        noise_subset = data_arg[:, -2] == noise_sample_idx
+        subset = signal_subset & noise_subset
+        if sum(subset) == 0:
+            continue
+
+        indices = np.where(subset)[0]
+        barcode_combo = "_".join([str(noise_sample_idx), str(signal_sample_idx)])
+        all_indices[np.where(subset)[0]] = counter
+        counter_to_barcode_combo[counter] = barcode_combo
+        counter += 1
+        noise_params = noise_params_dict[noise_sample_idx]
+        signal_params = signal_params_dict[signal_sample_idx]
+
+        # calculate probabilties for each hypothesis for each cell
+        data_subset = data[subset]
+        log_signal_signal_probs = np.log(
+            norm.pdf(
+                data_subset[:, signal_sample_idx],
+                *signal_params[:-2],
```

Comment:
btw I'm not sure this line does anything because you already specify loc and scale